File: api-code-examples/rust/docs-share.rs
use iroh::{base::node_addr::AddrInfoOptions, client::docs::ShareMode};

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    // Create in memory iroh node
    let node = iroh::node::Node::memory().enable_docs().spawn().await?;

    let node_id = node.net().node_id().await?;
    println!("Started Iroh node: {node_id}");

    let doc = node.docs().create().await?;
    println!("Created doc: {}", doc.id());

    let ticket = doc.share(ShareMode::Read, AddrInfoOptions::Id).await?;
    println!("Read-Access Ticket: {ticket}");

    let ticket = doc.share(ShareMode::Write, AddrInfoOptions::Id).await?;
    println!("Write-Access Ticket: {ticket}");

    Ok(())
}
// Output:
// Started Iroh node: rwgutd2wazt756h5awh6x576jejizai6w2l6ae5gxsv7tpswmopq
// Created doc: ijgvnoxfgkkgze3cyb7rwjywfqvfzy2pzilo2tn5wf64fgm7nxkq
// Read-Access Ticket: docafbe2vv24uzji3etmlah6gzhcywcuxhdj7fbn3knxwyx3quzt5w5kajarwgutd2wazt756h5awh6x576jejizai6w2l6ae5gxsv7tpswmopqcaifabfesmskyrlqbqfiahf4ivybeybxaaehhlf5maaaaaaaaaa2fhcvoajganyabbz2zplazdjnblb2x3kfyvlqcjqdoaaioowl2zctm6q34dlgzp6fk4
// Write-Access Ticket: docaaqo2dtt7udula5vvevfwjenux3tk3l3lfpefugeu6r6v55pwusvc5qbecgy2smpkydgp7xy7ucy727x7zerfdebd23jpyatu26kx6n6kzrz6aibauaeusjsjlcfoagavaa4xrcxaetag4aaq45mxvqaaaaaaaaadiu4kvybeybxaaehhlf5mdenfufmhk7nixcvoajganyabbz2zpleknt2dpqnm3f7yvlq

File: api-code-examples/nodejs/docs-share.js
import { AddrInfoOptions, Iroh, ShareMode } from '@number0/iroh'

// Setup Alice
const alice = await Iroh.memory()
const aliceId = await alice.net.nodeId()
console.log(`Started alice: ${aliceId}`)

// Setup Bob
const bob = await Iroh.memory()
const bobId = await bob.net.nodeId()
console.log(`Started bob: ${bobId}`)

// Alice creates a document
const aliceDoc = await alice.docs.create()
console.log(`Alice created doc: ${aliceDoc.id()}`)

// Create a ticket to share with Bob
const ticket = await aliceDoc.share(ShareMode.Read, AddrInfoOptions.RelayAndAddresses)

// Bob joins the doc from the ticket:
const bobDoc = await bob.docs.joinAndSubscribe(ticket, (err, progress) => {
  if (err) {
    console.log(`doc error: ${error}`)
    return
  }

  if (progress.insertLocal) { console.log(`insertLocal event`) }
  else if (progress.insertRemote) { console.log(`insertRemote event`) }
  else if (progress.neighborUp) { console.log(`neighborUp event`) }
  else if (progress.neighborDown) { console.log(`neighborDown event`) }
  else if (progress.contentReady) { console.log(`contentReady event`) }
  else if (progress.pendingContentReady) { console.log(`pendingContentReady event`) }
  else if (progress.syncFinished) { console.log(`syncFinished event`) }
  else { console.log(`unrecognized doc event: ${progress}`) }
})

await aliceDoc.closeMe()
await bobDoc.closeMe()